server/src/main/java/server/haengdong/domain/MemberActionFactory.java

```diff
@@ -0,0 +1,65 @@
+package server.haengdong.domain;
+
+import java.util.ArrayList;
+import java.util.List;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Component;
+import server.haengdong.application.request.MemberActionSaveAppRequest;
+import server.haengdong.application.request.MemberActionSaveListAppRequest;
+import server.haengdong.persistence.ActionRepository;
+
+@RequiredArgsConstructor
+@Component
+public class MemberActionFactory {
+
+    private final ActionRepository actionRepository;
+    private final MemberGroupIdProvider memberGroupIdProvider;
+
+    public List<MemberAction> createMemberActions(
+            MemberActionSaveListAppRequest request,
+            List<MemberAction> memberActions,
+            Event event
+    ) {
+        validateActions(request, memberActions);
+        Long memberGroupId = memberGroupIdProvider.createGroupId();
+        long lastSequence = getLastSequence(event);
+        List<MemberAction> memberActionList = new ArrayList<>();
+        List<MemberActionSaveAppRequest> actions = request.actions();
+        for (MemberActionSaveAppRequest appRequest : actions) {
+            Action action = new Action(event, ++lastSequence);
+            MemberAction memberAction = appRequest.toMemberAction(action, memberGroupId);
+            memberActionList.add(memberAction);
+        }
+        return memberActionList;
+    }
+
+    private void validateActions(MemberActionSaveListAppRequest request, List<MemberAction> memberActions) {
+        for (MemberActionSaveAppRequest action : request.actions()) {
+            validateAction(memberActions, action);
+        }
+    }
+
+    private void validateAction(List<MemberAction> memberActions, MemberActionSaveAppRequest action) {
+        if (!isAvailableAction(memberActions, action.name(), action.status())) {
+            throw new IllegalArgumentException();
+        }
+    }
+
+    private long getLastSequence(Event event) {
+        return actionRepository.findLastByEvent(event)
+                .map(Action::getSequence)
+                .orElse(0L);
+    }
+
+    private boolean isAvailableAction(List<MemberAction> actions, String name, String status) {
+        MemberActionStatus memberActionStatus = MemberActionStatus.of(status);
+        for (int i = actions.size() - 1; i >= 0; i--) {
+            MemberAction action = actions.get(i);
+            if (action.isSameName(name)) {
+                return action.isAvailable(memberActionStatus);
+            }
+        }
+
+        return MemberActionStatus.isMemberStatusIn(memberActionStatus);
```

Comment:
c: validateAction()을 호출하기 전 다음과 같이 memberActions에 대한 정렬을 수행하면 가독성이 더 좋을 것 같아요

```java
memberActions.sort(Comparator.comparing(MemberAction::getSequence).reversed());
```

```suggestion
        for (MemberAction action : actions) {
            if (action.isSameName(name)) {
                return action.isAvailable(memberActionStatus);
            }
        }
        return MemberActionStatus.IN == memberActionStatus;
```

Reply:
제안 감사합니다 :) 부분적으로 수용했는데 `reversed()`는 반영하지 않았어요. 

- 이전 로그를 기반으로 역추적하는 검증 로직을 작성했는데요. 
- 로그의 마지막 부분부터 반대로 스캔하는 방식이에요. 
- 이것이 검증 로직에 드러나야 할 것 같아요. 

코드를 수정하고 의견 부탁드려요!

Reply:
코드를 처음 본 사람이 validateAction 메서드 호출 시 로그를 반대로 정렬해서 줘야 한다는 사실을 파악하기는 어려울 테니, 부분적으로 수용한 현재 코드가 더 예측하기 쉽겠네요. 잘 수정된 것 같습니다!